sdk/provisioning/Azure.Provisioning/src/Primitives/ProvisioningConstruct.cs

```diff
@@ -0,0 +1,201 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using Azure.Provisioning.Expressions;
+
+namespace Azure.Provisioning.Primitives;
+
+/// <summary>
+/// A named Bicep entity, like a resource or parameter.
+/// </summary>
+public abstract class NamedProvisioningConstruct : ProvisioningConstruct
+{
+    /// <summary>
+    /// Gets the the Bicep name of the resource.  This can be used to refer to
+    /// the resource in expressions, but isn't the Azuree name of the resource.
+    /// </summary>
+    public string ResourceName { get; private set; }
+
+    public NamedProvisioningConstruct(string resourceName, ProvisioningContext? context = default)
+        : base(context)
+    {
+        ResourceName = resourceName;
+
+        // Named entities get added to the default infrastructure automatically
+        DefaultProvisioningContext.DefaultInfrastructure.Add(this);
+    }
+}
+
+public abstract class ProvisioningConstruct(ProvisioningContext? context = default) : Provisionable
```

Comment:
`ProvisioningConstruct` is next and adds the ability to have "fancy Bicep properties."  We'll get to that later, but just remember `ProvisioningConstruct` is where all the property goodness starts.

Most of the blander model types are strongly typed `ProvisioningConstruct`s.